File: src/main/java/com/github/pipding/locatesharedlibraryfunctionsplugin/SharedLibraryFunctionGoToDeclarationHandler.java
package com.github.pipding.locatesharedlibraryfunctionsplugin;

import com.intellij.codeInsight.navigation.actions.GotoDeclarationHandler;
import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.editor.Editor;
import com.intellij.psi.PsiElement;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public class SharedLibraryFunctionGoToDeclarationHandler implements GotoDeclarationHandler {
    @Override
    public PsiElement @Nullable [] getGotoDeclarationTargets(@Nullable PsiElement sourceElement, int offset, Editor editor) {

        PsiElement target;

        // Is the selected element a filename?
        target = SharedLibraryFunctionUtility.FindSharedLibraryScriptByElement(sourceElement);

        if (target != null) {
            return List.of(target).toArray(new PsiElement[0]);
        }

        target = SharedLibraryFunctionUtility.FindSharedLibraryFunctionByElement(sourceElement);

        if (target == null) {
            return new PsiElement[0];
        }
        else {
            return List.of(target).toArray(new PsiElement[0]);
        }
    }

    @Override
    public @Nullable @Nls(capitalization = Nls.Capitalization.Title) String getActionText(@NotNull DataContext context) {
        return GotoDeclarationHandler.super.getActionText(context);
    }
}

File: src/main/kotlin/com/github/pipding/locatesharedlibraryfunctionsplugin/SharedLibraryFunctionGoToDeclarationHandler.kt
//package com.github.pipding.locatesharedlibraryfunctionsplugin
//
//import com.intellij.codeInsight.navigation.actions.GotoDeclarationHandler
//import com.intellij.openapi.editor.Editor
//import com.intellij.psi.PsiElement
//import com.intellij.psi.util.parentOfType
//
//class SharedLibraryFunctionGoToDeclarationHandler : GotoDeclarationHandler {
//    override fun getGotoDeclarationTargets(sourceElement: PsiElement?, offset: Int, editor: Editor?): Array<PsiElement>? {
//        if (sourceElement == null) {
//            return null
//        }
//
//        val toml = findTomlFile(sourceElement, sourceElement.text)
//        if (toml != null) {
//            return arrayOf(toml)
//        }
//
//        return null
//    }
//
//    private fun findTomlFile(context: PsiElement, name: String): TomlFile? {
//        context.getRootGradleProjectPath()
//        val module = context.containingFile?.originalFile?.virtualFile?.let {
//            ProjectFileIndex.getInstance(context.project).getModuleForFile(it)
//        } ?: return null
//        val tomlPath = context.project.service<VersionCatalogsLocator>().getVersionCatalogsForModule(module)[name] ?: return null
//        val toml = VfsUtil.findFile(tomlPath, false) ?: return null
//        return PsiManager.getInstance(context.project).findFile(toml)?.castSafelyTo<TomlFile>()
//    }
//}